// devexpress-dashboard/model/format-rules/format-rules-common.js
/**
* DevExpress Dashboard (format-rules-common.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComplexValue = void 0;
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var ko = require("knockout");
var _parameters_helper_1 = require("../parameters/_parameters-helper");
var serializable_model_1 = require("../serializable-model");
var _format_rules_common_1 = require("./metadata/_format-rules-common");
var ComplexValue = (function (_super) {
    __extends(ComplexValue, _super);
    function ComplexValue(modelJson, serializer) {
        if (modelJson === void 0) { modelJson = {}; }
        if (serializer === void 0) { serializer = new analytics_utils_1.ModelSerializer(); }
        var _this = _super.call(this, modelJson, serializer) || this;
        if (!_this.isEmpty() && !_this.isInfinity) {
            _this._persistedValue(_parameters_helper_1.ParameterHelper.convertSingleValue(_this._persistedValue(), _this.type()));
        }
        _this.value = ko.pureComputed({
            read: function () { return _this._persistedValue(); },
            write: function (val) {
                if (val === _format_rules_common_1.negativeInfinity) {
                    _this.type('System.Double');
                }
                _this._persistedValue(val);
            }
        });
        return _this;
    }
    ComplexValue.prototype.isEmpty = function () {
        return !this._persistedValue || !this.type || !this.type();
    };
    Object.defineProperty(ComplexValue.prototype, "isInfinity", {
        get: function () {
            return this._persistedValue() == _format_rules_common_1.negativeInfinity;
        },
        enumerable: false,
        configurable: true
    });
    ComplexValue.prototype.getInfo = function () {
        return _format_rules_common_1.complexValueInfo;
    };
    ComplexValue.prototype.setValue = function (value, type) {
        this.value(value);
        this.type(type);
    };
    return ComplexValue;
}(serializable_model_1.SerializableModel));
exports.ComplexValue = ComplexValue;
