// devexpress-dashboard/model/format-rules/style-settings/color-style-settings.js
/**
* DevExpress Dashboard (color-style-settings.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorStyleSettings = void 0;
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var color_1 = require("../../color");
var _color_style_settings_1 = require("./metadata/_color-style-settings");
var style_settings_base_1 = require("./style-settings-base");
var ColorStyleSettings = (function (_super) {
    __extends(ColorStyleSettings, _super);
    function ColorStyleSettings(modelJson, serializer) {
        if (modelJson === void 0) { modelJson = {}; }
        if (serializer === void 0) { serializer = new analytics_utils_1.ModelSerializer(); }
        return _super.call(this, modelJson, serializer) || this;
    }
    ColorStyleSettings.prototype.getInfo = function () {
        return _color_style_settings_1.colorStyleSettingsSerializationsInfo;
    };
    ColorStyleSettings.prototype.equals = function (style) {
        return (style instanceof ColorStyleSettings) &&
            this.predefinedColor() === style.predefinedColor() &&
            color_1.Color.equals(this.color(), style.color());
    };
    ColorStyleSettings.prototype.clone = function () {
        var style = new ColorStyleSettings({});
        style.predefinedColor(this.predefinedColor());
        style.color(this.color());
        return style;
    };
    ColorStyleSettings.prototype.setSpecificType = function (type) {
        this.predefinedColor(type);
    };
    ColorStyleSettings.prototype.init = function () {
        this.predefinedColor('Red');
    };
    return ColorStyleSettings;
}(style_settings_base_1.StyleSettingsBase));
exports.ColorStyleSettings = ColorStyleSettings;
