// devexpress-dashboard/designer/extensions/_available-data-sources-view-model.js
/**
* DevExpress Dashboard (_available-data-sources-view-model.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvailableDataSourcesViewModel = void 0;
var AvailableDataSourcesViewModel = (function () {
    function AvailableDataSourcesViewModel(_dataSources, selectedDataSources, uiState, errorState, _showCreateDataSourceWizardDelegate) {
        var _this = this;
        this._dataSources = _dataSources;
        this.selectedDataSources = selectedDataSources;
        this.uiState = uiState;
        this.errorState = errorState;
        this._showCreateDataSourceWizardDelegate = _showCreateDataSourceWizardDelegate;
        this.showCreateDataSourceWizard = function (federationSources) {
            if (_this.canCreateDataSourceWizard) {
                _this._showCreateDataSourceWizardDelegate()(federationSources);
            }
        };
    }
    AvailableDataSourcesViewModel.prototype.getDataSources = function (dataSourcesFilter) {
        if (dataSourcesFilter === void 0) { dataSourcesFilter = (function () { return true; }); }
        return this._dataSources().filter(dataSourcesFilter);
    };
    Object.defineProperty(AvailableDataSourcesViewModel.prototype, "canCreateDataSourceWizard", {
        get: function () {
            return !!this._showCreateDataSourceWizardDelegate();
        },
        enumerable: false,
        configurable: true
    });
    return AvailableDataSourcesViewModel;
}());
exports.AvailableDataSourcesViewModel = AvailableDataSourcesViewModel;
