// devexpress-dashboard/designer/extensions/available-data-sources-extension.js
/**
* DevExpress Dashboard (available-data-sources-extension.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvailableDataSourcesExtension = void 0;
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var ko = require("knockout");
var control_options_1 = require("../../common/control-options");
var notificator_1 = require("../../common/notification-controller/notificator");
var _default_1 = require("../../data/localization/_default");
var dashboard_1 = require("../../model/dashboard");
var disposable_object_1 = require("../../model/disposable-object");
var _helper_classes_1 = require("../../model/internal/_helper-classes");
var _knockout_utils_1 = require("../../model/internal/_knockout-utils");
var _obsolete_helper_1 = require("../../model/internal/_obsolete-helper");
var _available_data_sources_view_model_1 = require("./_available-data-sources-view-model");
var name = 'available-data-sources';
var nameAlias = 'availableDataSources';
var AvailableDataSourcesExtension = (function (_super) {
    __extends(AvailableDataSourcesExtension, _super);
    function AvailableDataSourcesExtension(dashboardControl) {
        var _this = _super.call(this) || this;
        _this.dashboardControl = dashboardControl;
        _this.name = name;
        _this.templateName = 'dx-dashboard-datasource-available-datasources';
        _this.selectedDataSources = ko.observableArray();
        _this.dataSources = ko.observableArray();
        _this._errorState = ko.observable(null);
        _this._uiState = ko.observable('empty');
        var showCreateDataSourceWizardDelegate = ko.computed(function () {
            var dataSourceWizardExtension = (_this.dashboardControl.findExtension('data-source-wizard'));
            if (dataSourceWizardExtension) {
                return function (federationSources) {
                    var createdDataSourcePromise = federationSources ? dataSourceWizardExtension._showDataSourceCreatingDialog(federationSources) : dataSourceWizardExtension.showDataSourceCreatingDialog();
                    createdDataSourcePromise.done(function (dataSource) {
                        _helper_classes_1.NameGenerator.validateName(dataSource, _this.dataSources(), 'name', 1, true);
                        _this.dataSources.push(dataSource);
                    });
                };
            }
            else {
                return null;
            }
        });
        _this.toDispose(showCreateDataSourceWizardDelegate);
        var isInitialized = false;
        var uiStateComputed = ko.computed(function () {
            if (!isInitialized) {
                _this.loadAvailableDataSources();
                isInitialized = true;
            }
            return _this._uiState();
        }, _this, {
            deferEvaluation: true
        });
        _this.toDispose(uiStateComputed);
        _this.viewModel = new _available_data_sources_view_model_1.AvailableDataSourcesViewModel(_this.dataSources, _this.selectedDataSources, uiStateComputed, _this._errorState, showCreateDataSourceWizardDelegate);
        _obsolete_helper_1.defineObsoleteMethod({
            target: _this,
            memberName: 'loadAvaliableDataSources',
            oldMemberDisplayName: 'AvailableDataSourcesExtension.loadAvaliableDataSources',
            newMemberDisplayName: 'AvailableDataSourcesExtension.loadAvailableDataSources',
            action: function () { return _this.loadAvailableDataSources(); }
        });
        return _this;
    }
    AvailableDataSourcesExtension.prototype.start = function () {
        var _this = this;
        if (this.dataSources().length > 0) {
            this.selectedDataSources([this.dataSources()[0]]);
        }
        _knockout_utils_1.subscribeArrayChange(this.dataSources, {
            added: function (item) { return _this.selectedDataSources([item]); }
        });
    };
    AvailableDataSourcesExtension.prototype.stop = function () {
        this._errorState(null);
    };
    AvailableDataSourcesExtension.prototype.loadAvailableDataSources = function () {
        var _this = this;
        if (this.dashboardControl._endpointCollection.dataSourceUrls) {
            this._uiState('loading');
            this.dashboardControl.remoteService.getFromServer(this.dashboardControl._endpointCollection.dataSourceUrls.GetDataSourcesAction)
                .then(function (result) {
                var dataSources = analytics_utils_1.deserializeArray(result, function (item) { return dashboard_1.Dashboard._createDataSource(item, new analytics_utils_1.ModelSerializer()); })();
                dataSources.forEach(function (dataSource) {
                    if (!dataSource.name()) {
                        dataSource.name(_helper_classes_1.NameGenerator.generateName(_default_1.getLocalizationById('DashboardStringId.DefaultDataSourceName') + ' ', dataSources, 'name', 1));
                    }
                });
                _this.dataSources(dataSources);
                _this._uiState('live');
            }, function (errorInfo) {
                var errorDetail = notificator_1.NotificationController._getDetailedErrorMessage(errorInfo);
                _this._errorState({
                    title: _default_1.getLocalizationById('DashboardWebStringId.DataSources.AvailableDataSourcesError'),
                    detail: errorDetail
                });
                _this._uiState('error');
            });
        }
    };
    return AvailableDataSourcesExtension;
}(disposable_object_1.DisposableObject));
exports.AvailableDataSourcesExtension = AvailableDataSourcesExtension;
control_options_1.designerExtensions[name] = function (dashboardControl, options) { return new AvailableDataSourcesExtension(dashboardControl); };
control_options_1.extensionNameMap[name] = nameAlias;
