// devexpress-dashboard/model/internal/_сlient-filter-manager.js
/**
* DevExpress Dashboard (_сlient-filter-manager.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientFilterManager = void 0;
var ClientFilterManager = (function () {
    function ClientFilterManager(_updateTotals) {
        var _this = this;
        this._updateTotals = _updateTotals;
        _updateTotals.subscribe(function (value) {
            if (!value)
                _this.clearFilterParams();
        });
    }
    ClientFilterManager.prototype.getFilterParams = function () {
        return this._filterParams;
    };
    ClientFilterManager.prototype.setFilterParams = function (clientFilter) {
        if (this._updateTotals()) {
            this._filterParams = clientFilter;
        }
    };
    ClientFilterManager.prototype.clearFilterParams = function () {
        this._filterParams = undefined;
    };
    ClientFilterManager.prototype.setFilterState = function (filterValue) {
        this._filterState = filterValue;
    };
    ClientFilterManager.prototype.getFilterState = function () {
        return this._filterState;
    };
    return ClientFilterManager;
}());
exports.ClientFilterManager = ClientFilterManager;
