// devexpress-dashboard/designer/filtering/simple-filter-editor/_simple-filter-editor-view-model.js
/**
* DevExpress Dashboard (_simple-filter-editor-view-model.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleFilterExpressionEditorViewModel = void 0;
var ko = require("knockout");
var _simple_filter_editor_1 = require("./_simple-filter-editor");
var SimpleFilterExpressionEditorViewModel = (function () {
    function SimpleFilterExpressionEditorViewModel(options) {
        this.options = options;
        this.editorTemplate = 'dx-simple-filter-popup';
        this.editor = new _simple_filter_editor_1.SimpleFilterEditor(options.dashboardItem, options.dataSourceBrowser);
        this.criteriaString = ko.computed(function () { return options.dashboardItem.filterString(); });
    }
    SimpleFilterExpressionEditorViewModel.prototype.edit = function () {
        this.editor.popupVisible(true);
    };
    SimpleFilterExpressionEditorViewModel.prototype.remove = function () {
        this.options.dashboardItem.filterString(null);
    };
    return SimpleFilterExpressionEditorViewModel;
}());
exports.SimpleFilterExpressionEditorViewModel = SimpleFilterExpressionEditorViewModel;
