// devexpress-dashboard/designer/text-box-item-editor-extension/_text-box-item-editor-view-model.js
/**
* DevExpress Dashboard (_text-box-item-editor-view-model.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RichEditExtensionViewModel = exports.ValueAccessor = void 0;
var ko = require("knockout");
var _default_1 = require("../../data/localization/_default");
var _confirm_dialog_1 = require("../confirm-dialog/_confirm-dialog");
var ValueAccessor = (function () {
    function ValueAccessor() {
    }
    return ValueAccessor;
}());
exports.ValueAccessor = ValueAccessor;
var RichEditExtensionViewModel = (function () {
    function RichEditExtensionViewModel() {
        var _this = this;
        this._isCloseConfirmed = false;
        this.richEditorOptions = ko.observable(null);
        this.close = function () {
            _this.popup.hide();
        };
        this.saveAndClose = function () {
            var richEditorOptions = _this.richEditorOptions();
            richEditorOptions && richEditorOptions.forceSave && richEditorOptions.forceSave();
            _this.close();
        };
        this.onHidden = function () {
            _this.richEditorOptions(null);
            _this._isCloseConfirmed = false;
        };
        this.onHiding = function (args) {
            var richEditorOptions = _this.richEditorOptions();
            var documentModified = richEditorOptions && richEditorOptions.documentModified && richEditorOptions.documentModified();
            if (documentModified && !_this._isCloseConfirmed) {
                args.cancel = true;
                _this.confirmDialogViewModel
                    .confirm(_default_1.getLocalizationById('DashboardWebStringId.Dialog.ConfirmSaving'), _default_1.getLocalizationById('DashboardWebStringId.SaveConfirmationDialogMessage') + '<br/>' + _default_1.getLocalizationById('DashboardWebStringId.SaveChangesDialogMessage'), _default_1.getLocalizationById('DashboardWebStringId.Dialog.Save'), _default_1.getLocalizationById('DashboardWebStringId.Dialog.DoNotSave'))
                    .done(function (result) {
                    _this._isCloseConfirmed = true;
                    if (result) {
                        _this.saveAndClose();
                    }
                    else {
                        _this.close();
                    }
                });
            }
        };
        this.onInitialized = function (args) {
            _this.popup = args.component;
        };
        this.confirmDialogViewModel = new _confirm_dialog_1.ConfirmDialogViewModel();
    }
    RichEditExtensionViewModel.prototype.show = function (options, docVariables, dashboardItemWidth) {
        var _this = this;
        if (!this.richEditorOptions()) {
            this.popup.show().then(function () {
                _this.richEditorOptions({
                    initialDocument: options.getValue(),
                    docVariables: docVariables,
                    onSaving: function (document) {
                        options.setValue(document);
                    },
                    dashboardItemWidth: dashboardItemWidth
                });
            });
        }
    };
    return RichEditExtensionViewModel;
}());
exports.RichEditExtensionViewModel = RichEditExtensionViewModel;
