// devexpress-dashboard/model/items/_limit-data-state.js
/**
* DevExpress Dashboard (_limit-data-state.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimitDataState = void 0;
var LimitDataState = (function () {
    function LimitDataState() {
        this._visible = false;
        this._enabled = true;
    }
    Object.defineProperty(LimitDataState.prototype, "enabled", {
        get: function () {
            return this._enabled;
        },
        enumerable: false,
        configurable: true
    });
    LimitDataState.prototype.setReduced = function () {
        this._visible = true;
        this._enabled = true;
    };
    LimitDataState.prototype.getViewModel = function () {
        return {
            isReduceMode: this._visible,
            isReduced: this._enabled,
        };
    };
    LimitDataState.prototype.toggle = function () {
        this._enabled = !this._enabled;
    };
    LimitDataState.prototype.reset = function () {
        this._visible = false;
        this._enabled = true;
    };
    return LimitDataState;
}());
exports.LimitDataState = LimitDataState;
