// devexpress-dashboard/designer/items/surfaces/_delta-numeric-format-surface.js
/**
* DevExpress Dashboard (_delta-numeric-format-surface.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeltaNumericFormatSurface = void 0;
var ko = require("knockout");
var _accordion_tab_1 = require("../../properties-controller/_accordion-tab");
var _shared_composers_1 = require("../properties-composers/_shared-composers");
var DeltaNumericFormatSurface = (function () {
    function DeltaNumericFormatSurface(model, propertiesController) {
        this.model = model;
        this.propertiesController = propertiesController;
        this._disposables = [];
        this.propertiesTabs = ko.observableArray([]);
    }
    DeltaNumericFormatSurface.prototype.updatePropertiesTabs = function () {
        this.propertiesTabs([new _accordion_tab_1.AccordionTab(_accordion_tab_1.KnownTabs.NumericFormat, this.model.title, _shared_composers_1.SharedComposers.getNumericFormatWrapper(this.model.numericFormat))
        ]);
        this.propertiesController.secondaryModel({
            displayText: this.model.title,
            data: this
        });
    };
    DeltaNumericFormatSurface.prototype.startEditing = function (args) {
        args.createImmediately = false;
        this.updatePropertiesTabs();
    };
    DeltaNumericFormatSurface.prototype.dispose = function () {
        this._disposables.forEach(function (d) {
            d.dispose();
        });
    };
    return DeltaNumericFormatSurface;
}());
exports.DeltaNumericFormatSurface = DeltaNumericFormatSurface;
