// devexpress-dashboard/viewer-parts/widgets/caption-toolbar/_hidden-caption-toolbar.js
/**
* DevExpress Dashboard (_hidden-caption-toolbar.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HiddenCaptionToolbar = void 0;
var HiddenCaptionToolbar = (function () {
    function HiddenCaptionToolbar() {
        this.element = document.createElement('div');
        this.disabled = false;
    }
    HiddenCaptionToolbar.prototype.calcHeight = function (options) {
        return 0;
    };
    HiddenCaptionToolbar.prototype.calcMinWidth = function (options) {
        return 0;
    };
    HiddenCaptionToolbar.prototype.update = function (options) {
        return false;
    };
    HiddenCaptionToolbar.prototype.onResize = function () {
    };
    HiddenCaptionToolbar.prototype.dispose = function () {
    };
    return HiddenCaptionToolbar;
}());
exports.HiddenCaptionToolbar = HiddenCaptionToolbar;
