// devexpress-dashboard/designer/data-source-wizard/wizards/_data-source-wizard.js
/**
* DevExpress Dashboard (_data-source-wizard.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDashboardDataSourceWizard = exports.DashboardDataSourceWizardState = exports.DashboardDataSourceWizard = exports.DashboardDataSourceWizardBase = exports.EditQueryWizardIterator = exports.CreateNewDataSourceWizardIterator = void 0;
var analytics_wizard_1 = require("@devexpress/analytics-core/analytics-wizard");
var $ = require("jquery");
var ko = require("knockout");
var _default_1 = require("../../../data/localization/_default");
var data_source_wizard_model_1 = require("../models/data-source-wizard-model");
var _data_source_wizard_model_1 = require("../models/_data-source-wizard-model");
var choose_data_source_type_page_1 = require("../pages/choose-data-source-type-page");
var choose_olap_connection_string_page_1 = require("../pages/choose-olap-connection-string-page");
var federated_query_configure_page_1 = require("../pages/federated-query-configure-page");
var _choose_json_schema_page_1 = require("../pages/_choose-json-schema-page");
var _create_query_page_1 = require("../pages/_create-query-page");
var _parameters_item_provider_1 = require("../pages/_parameters-item-provider");
var _helpers_1 = require("../_helpers");
var _data_source_wizard_page_iterator_1 = require("./_data-source-wizard-page-iterator");
var CreateNewDataSourceWizardIterator = (function (_super) {
    __extends(CreateNewDataSourceWizardIterator, _super);
    function CreateNewDataSourceWizardIterator(factory, stateManager) {
        return _super.call(this, factory, stateManager) || this;
    }
    return CreateNewDataSourceWizardIterator;
}(_data_source_wizard_page_iterator_1.DataSourceWizardPageIteratorBase));
exports.CreateNewDataSourceWizardIterator = CreateNewDataSourceWizardIterator;
var EditQueryWizardIterator = (function (_super) {
    __extends(EditQueryWizardIterator, _super);
    function EditQueryWizardIterator(factory, stateManager) {
        return _super.call(this, factory, stateManager) || this;
    }
    EditQueryWizardIterator.prototype.getInitialPage = function () {
        return analytics_wizard_1.SqlDataSourceWizardPageId.ConfigureQueryPage;
    };
    return EditQueryWizardIterator;
}(_data_source_wizard_page_iterator_1.DataSourceWizardPageIteratorBase));
exports.EditQueryWizardIterator = EditQueryWizardIterator;
var DashboardDataSourceWizardBase = (function (_super) {
    __extends(DashboardDataSourceWizardBase, _super);
    function DashboardDataSourceWizardBase(factory) {
        return _super.call(this, factory) || this;
    }
    return DashboardDataSourceWizardBase;
}(analytics_wizard_1.PopupWizard));
exports.DashboardDataSourceWizardBase = DashboardDataSourceWizardBase;
var DashboardDataSourceWizard = (function (_super) {
    __extends(DashboardDataSourceWizard, _super);
    function DashboardDataSourceWizard(factory, options) {
        var _this = _super.call(this, factory) || this;
        _this.title = _default_1.getLocalizationById('DashboardWebStringId.DataSources.DashboardDataSourceWizard');
        _this._container = function (element) { return $(element).closest('.dx-dashboard-widget-container'); };
        _this._extendCssClass = 'dxrd-sqldatasource-wizard';
        _this._options = options;
        return _this;
    }
    Object.defineProperty(DashboardDataSourceWizard.prototype, "options", {
        get: function () {
            return this._options;
        },
        enumerable: false,
        configurable: true
    });
    return DashboardDataSourceWizard;
}(DashboardDataSourceWizardBase));
exports.DashboardDataSourceWizard = DashboardDataSourceWizard;
var DashboardDataSourceWizardState = (function () {
    function DashboardDataSourceWizardState() {
        this.sqlDataSourceWizard = {};
        this.jsonDataSourceWizard = {};
        this.olapDataSourceWizard = {};
        this.objectDataSourceWizard = {};
        this.federationDataSourceWizard = {};
    }
    Object.defineProperty(DashboardDataSourceWizardState.prototype, "dashboardDataSourceType", {
        get: function () {
            return this.dataSourceType && data_source_wizard_model_1.ToDashboardDataSourceType(this.dataSourceType) || undefined;
        },
        set: function (value) {
            this.dataSourceType = data_source_wizard_model_1.ToDataSourceTypeNumber(value);
        },
        enumerable: false,
        configurable: true
    });
    return DashboardDataSourceWizardState;
}());
exports.DashboardDataSourceWizardState = DashboardDataSourceWizardState;
function createDashboardDataSourceWizard(requestWrapper, parameters, disableCustomSql, allowCreateNewJsonConnection, wizardSettings, dashboardConnectionStrings, dataSources, dataSourceBrowser, customTemplates) {
    var factory = new analytics_wizard_1.PageFactory();
    var wizardOptions = new _data_source_wizard_model_1.DashboardDataSourceWizardOptions();
    wizardOptions.callbacks = _create_query_page_1.createQueryPageCallback(requestWrapper, parameters, dataSourceBrowser);
    wizardOptions.disableCustomSql = disableCustomSql;
    wizardOptions.requestWrapper = requestWrapper;
    wizardOptions.wizardSettings = _helpers_1.createWizardSettings(wizardSettings);
    wizardOptions.allowCreateNewJsonConnection = allowCreateNewJsonConnection || false;
    wizardOptions.connectionStrings = dashboardConnectionStrings;
    wizardOptions.dataSources = ko.pureComputed(function () { return dataSources; });
    choose_data_source_type_page_1._registerChooseDataSourceTypePage(factory, wizardOptions, customTemplates);
    analytics_wizard_1._registerChooseSqlConnectionPage(factory, dashboardConnectionStrings.sql);
    choose_olap_connection_string_page_1._registerOlapConnectionStringsPage(factory, dashboardConnectionStrings.olap);
    analytics_wizard_1._registerChooseJsonConnectionPage(factory, wizardOptions);
    analytics_wizard_1._registerConfigureQueryPage(factory, wizardOptions);
    analytics_wizard_1._registerConfigureParametersPage(factory, requestWrapper);
    analytics_wizard_1._registerChooseJsonSourcePage(factory, requestWrapper, function () { return new _parameters_item_provider_1.ParametersItemProvider(parameters()); });
    _choose_json_schema_page_1._registerChooseJsonSchemaPage(factory, requestWrapper);
    federated_query_configure_page_1._registerFederatedQueryConfigurePage(factory, wizardOptions);
    return new DashboardDataSourceWizard(factory, wizardOptions);
}
exports.createDashboardDataSourceWizard = createDashboardDataSourceWizard;
