// devexpress-dashboard/designer/data-source-wizard/pages/_create-query-page.js
/**
* DevExpress Dashboard (_create-query-page.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createQueryPageCallback = void 0;
var _helpers_1 = require("../_helpers");
var _parameters_item_provider_1 = require("./_parameters-item-provider");
function createQueryPageCallback(requestWrapper, parameters, dataSourceBrowser, customQueriesPreset) {
    if (customQueriesPreset === void 0) { customQueriesPreset = undefined; }
    return {
        customQueriesPreset: customQueriesPreset,
        customizeQBInitData: function (initData) {
            initData.parametersItemsProvider = new _parameters_item_provider_1.ParametersItemProvider(parameters());
            initData.requestWrapper = requestWrapper;
            return initData;
        },
        selectStatement: function (connection, query) { return requestWrapper.getSelectStatement(connection, query); },
        getItemsProviderCallback: function () { return new _parameters_item_provider_1.ParametersItemProvider(parameters()); },
        fieldListsCallback: function (pathRequest, dataSourceInfo) { return _helpers_1.getFederationFieldList(pathRequest, dataSourceBrowser); },
    };
}
exports.createQueryPageCallback = createQueryPageCallback;
