// devexpress-dashboard/designer/data-source-wizard/pages/_parameters-item-provider.js
/**
* DevExpress Dashboard (_parameters-item-provider.js)
* Version:  22.1.3
* Build date: Jun 13, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParametersItemProvider = void 0;
var _jquery_helpers_1 = require("../../../data/_jquery-helpers");
var data_item_1 = require("../../../model/data-item/data-item");
var ParametersItemProvider = (function () {
    function ParametersItemProvider(dashboardParameters) {
        if (dashboardParameters === void 0) { dashboardParameters = []; }
        this.dashboardParameters = dashboardParameters;
    }
    ParametersItemProvider.prototype.getItems = function (pathRequest) {
        var items = [], fullPath = pathRequest.fullPath && pathRequest.fullPath.toLowerCase() || '';
        if (fullPath === '') {
            items = [{
                    displayName: 'Parameters',
                    name: 'Parameters',
                    isList: true,
                    specifics: 'parameters'
                }];
        }
        else if (fullPath === 'parameters') {
            items = this.dashboardParameters.map(function (parameter) {
                return {
                    displayName: parameter.name(),
                    name: parameter.name(),
                    isList: false,
                    specifics: data_item_1.DataItem.typesMap[parameter.type().replace('System.', '')]
                };
            });
        }
        return _jquery_helpers_1.createJQueryDeferred().resolve(items).promise();
    };
    return ParametersItemProvider;
}());
exports.ParametersItemProvider = ParametersItemProvider;
